### ISSUE TITLE
updateInteractions() is in the wrong place -- should it be immediately after updateFreeState?

### ISSUE THREAD

**Issue by radarsat1** [siconos/siconos#103]
This mistake was due to my incomplete understanding of the integration process, which I hope I have improved now.  I think this could be related to #102 as well as to issues with objects falling through the floor and to the appearance of "ghost" forces that cause objects to suddenly fly away under extreme conditions of friction and gravity.

The summarize, when implementing the InteractionManager, I took the liberty of putting together the steps of updating the collision positions from DS->q() and calculating contacts/gap distance, and wrongly put this at the beginning of advanceToEvent(), outside the Newton loop.  I did not understand properly the implication of updateWorldFromDS() being called in-between updateState() and updateOutput().

Moving the call to updateInteraction() from advanceToEvent() to where updateWorldFromDS() is called in TimeStepping::update() (commit 5fb187a10abede36b59d1469571ef2d006c4793b) seems to greatly improve convergence for the problems reported in #102.

## However I still think this might be wrong!

It still does not solve falling-through-the-floor issues.  (Try running tetra.py with h=0.001, it penetrates *worse* than with h=0.005)

So here is a question:
* gap distance is calculated in computeh().
* In the Bullet implementation, BulletR::computeh() simply reports distances based on collisions
* In the Newton loop, collisions are calculated based on state after updateState(), called from TimeStepping::update(), *after* computeOneStepNSProblem()

However, doesn't the solver need to consider the gap distance of the *free* state?

If so, then updateFreeState() needs to call updatePosition(), and updateInteractions() must be called *before* computeOneStepNSProblem().

I am not sure if this is correct.

But if so, it requires updating DS->q from the free velocity, updating collisions (updateInteractions), and then calling the solver, restoring DS->q and updating from the updated velocity.  I have tried but not yet succeeded in testing this.

**Comment by radarsat1** [comment]
Note that I would prefer not to add/remove interactions in the middle of the Newton loop!  So I am not 100% sure my change is correct.

However, at least with Bullet, it seems impossible or very difficult to separate the calculation of gap distance from addition/removal of contact points.

**Comment by radarsat1** [comment]
Ok, I think I just answered my own question, the free velocity is used to transform yForNSsolver by the jacobian jachqT during computeFreeOutput, which is called during computeq() to prepare input for the solver.

The particular jachqT calculation used in NewtonEulerFrom3DLocalFrameR (used with Bullet) is responsible for projecting the free velocity onto the velocity impact constraint.

So basically NewtonEulerFrom3DLocalFrameR only needs the initial gap distance and the body's free velocity, which it does receive through the Xfree workspace variable.  It doesn't need to update the gap distance according to the free velocity, since it is applying a constraint against admissible velocities.

**Comment by vacary** [comment]
I didn't get time to answer but again you find the solution.

The distance is only used for activating interaction (indexset0 --> indexset1).

If the body moves and worse rotates quickly we need to update jachqT in the loop of Newton. The other problem  is the fact that T(q) is not updated correctly. If we want to update we need to update the object into bullet before computing JacqT

**Comment by radarsat1** [comment]
I am improving my understanding of Siconos this week!

Actually I just discovered that with the fix in 5fb187a10abede36b59d1469571ef2d006c4793b, Gazebo also runs much, much faster.  So I think it is a big improvement.  I am only worried that it is not the right place to add/remove Interactions, so I will continue thinking about that.
The wheels.py example does also work correctly now, both versions.  Although it still produces some local solver errors in some cases, but they do not diverge.

Your explanations of updating jachqT now make sense to me, so I guess I have learned something ;)
